functions/snippets/src/test/java/HelloBackgroundTest.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import static org.mockito.Mockito.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.cloud.functions.HttpRequest;
+import com.google.cloud.functions.HttpResponse;
+import com.google.gson.Gson;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.Optional;
+import java.util.logging.Logger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({Logger.class, HelloBackground.class})
+public class HelloBackgroundTest {
+  @Mock private static Logger loggerInstance;
```

Comment:
Mocking a Logger seems kind of complicated to me. `com.google.common.testing.TestLogHandler` seems like a better way to go.

Reply:
Ideally, we should remove the mocking all together and just verify the response of the function. Mocking introduces additional complexity that I don't think is necessary for these samples.

Reply:
IIRC the reason for mocking here is that Stackdriver is a) eventually consistent, and b) achieving  consistency (and finding the required log entries) can take up to several minutes.

Reply:
I'm not sure I follow. If you add a TestLogHandler then you can inspect it immediately after calling the code that is supposed to have logged. It just saves up log records locally so Stackdriver is not involved.

Reply:
Ah, gotcha - you were talking about a local log handler.

I did some research on this, and it seems like we'd have to add additional code to the sample that's only there for our testing purposes. @kurtisvg is that something you're OK with, idiomaticity-wise?

In any case, I also found out that the `verify` assertions in Mockito **were not running properly.** (They silently fail, which causes tests to pass **even if values are incorrect!** 🤢)

Reply:
1.) One, I don't understand why we need to use a logger. What is the downside to just writing to response? This seems easier to test and verify. 

2.) This snippet seems to have diverged from the intent. The Go version of this sample is about handling a PubSub message - should this be the same?

Reply:
Some samples run in the background, and will (AFAIK) need to log input to Stackdriver.

We can try to minimize the frequency of this code, but IMO we need to show users how to log at different log levels _somewhere_.

Reply:
It should not be necessary to modify the sample for what I suggest to work. The sample does

`Logger.getLogger(HelloBackground.class.getName())`

to define its logger, which is completely idiomatic. All we need to do in the test is to do the same thing in order to get the same `Logger` instead, to which we can then add a `TestLogHander`.

I don't see anything about this test that would involve Stackdriver in any way. Sure, it should be involved if one of these samples is deployed to GCF, but the test doesn't do that.

Reply:
Hmm - I was having trouble retrieving log entries from `TestLogHandler`, but apparently declaring instances of the latter as **class**-level variables (instead of **function**-level ones) solved it. 🤷‍♂